File: 50. Pow(x, n)/main.cpp
#include <cmath>
#include <limits>

using namespace std;

class Solution {
public:
    double myPow(double x, int n) {
        return pow(x, n);
    }
};

class Solution {
public:
    double myPow(double x, int n) {
        if (fabs(x) < numeric_limits<double>::epsilon()) {
            return 0.0;
        }

        if (n == 0) {
            return 1.0;
        }

        if (n < 0) {
            return 1 / (x * myPow(x, -1 - n));
        }

        double power = myPow(x, n / 2);

        if (n % 2 == 0) {
            return power * power;
        }

        return x * power * power;
    }
};

class Solution {
public:
    double myPow(double x, int n) {
        if (fabs(x) < numeric_limits<double>::epsilon()) {
            return 0.0;
        }

        if (n == 0) {
            return 1.0;
        }

        long long num = n;
        if (num < 0) {
            x = 1 / x;
            num = -num;
        }

        double result = 1.0;
        while (num > 0) {
            if (num % 2 == 1) {
                result *= x;
            }

            x *= x;
            num /= 2;
        }

        return result;
    }
};

File: 50. Pow(x, n)/main.py
import math
import sys


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if math.fabs(x) < sys.float_info.epsilon:
            return 0.0

        if n == 0:
            return 1.0

        if n < 0:
            return 1 / (x * self.myPow(x, -1 - n))

        power = self.myPow(x, n // 2)

        if n % 2 == 0:
            return power * power

        return x * power * power


class Solution:
    def myPow(self, x: float, n: int) -> float:
        return math.pow(x, n)


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if math.fabs(x) < sys.float_info.epsilon:
            return 0.0

        if n == 0:
            return 1.0

        if n < 0:
            x = 1 / x
            n = -n

        result = 1.0

        while n > 0:
            if n % 2 == 1:
                result *= x

            x *= x
            n //= 2

        return result
